### PR TITLE
[SPARK-8174][SPARK-8175][SQL]function unix_timestamp, from_unixtime

### PR DESCRIPTION
**Pull request by adrian-wang**
unix_timestamp(): long
Gets current Unix timestamp in seconds.

unix_timestamp(string|date): long
Converts time string in format yyyy-MM-dd HH:mm:ss to Unix timestamp (in seconds), using the default timezone and the default locale, return null if fail: unix_timestamp('2009-03-20 11:30:01') = 1237573801

unix_timestamp(string date, string pattern): long
Convert time string with given pattern (see [http://docs.oracle.com/javase/tutorial/i18n/format/simpleDateFormat.html]) to Unix time stamp (in seconds), return null if fail: unix_timestamp('2009-03-20', 'yyyy-MM-dd') = 1237532400.

from_unixtime(bigint unixtime[, string format]): string
Converts the number of seconds from unix epoch (1970-01-01 00:00:00 UTC) to a string representing the timestamp of that moment in the current system time zone in the format of "1970-01-01 00:00:00".

Jira:
https://issues.apache.org/jira/browse/SPARK-8174
https://issues.apache.org/jira/browse/SPARK-8175

### CODE REVIEW

**Comment by adrian-wang** [comment]
retest this please.

**Comment by adrian-wang** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala [hunk 1]
```diff
@@ -247,14 +248,112 @@ case class DateFormatClass(left: Expression, right: Expression) extends BinaryEx
...
-    UTF8String.fromString(sdf.format(new Date(timestamp.asInstanceOf[Long] / 1000)))
+    UTF8String.fromString(sdf.format(new java.util.Date(timestamp.asInstanceOf[Long] / 1000)))
   }
 
   override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
     val sdf = classOf[SimpleDateFormat].getName
     defineCodeGen(ctx, ev, (timestamp, format) => {
       s"""UTF8String.fromString((new $sdf($format.toString()))
-          .format(new java.sql.Date($timestamp / 1000)))"""
+          .format(new java.sql.Timestamp($timestamp / 1000)))"""
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala]: why is this changed to Timestamp?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala [hunk 2]
```diff
@@ -247,14 +248,112 @@ case class DateFormatClass(left: Expression, right: Expression) extends BinaryEx
...
+/**
+ * Convert time string with given pattern
+ * (see [http://docs.oracle.com/javase/tutorial/i18n/format/simpleDateFormat.html])
+ * to Unix time stamp (in seconds), return null if fail.
+ * If the second parameter is missing, use "yyyy-MM-dd HH:mm:ss".
+ * If no parameters provided, the first parameter will be current_timestamp.
+ * If the first parameter is a Date or Timestamp instead of String, we will ignore the
+ * second parameter.
+ */
+case class UnixTimestamp(nodes: Seq[Expression])
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala]: This function would be simpler to just put two parameters, and add a ctor that takes in only one expression, and sets the 2nd parameter to Literal("yyyy-MM-dd HH:mm:ss").

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 3]
```diff
@@ -2419,6 +2419,44 @@ object functions {
...
+  def unix_timestamp(s: Column): Column = UnixTimestamp(Seq(s.expr))
+
+  /**
+   * Convert time string with given pattern
+   * (see [http://docs.oracle.com/javase/tutorial/i18n/format/simpleDateFormat.html])
+   * to Unix time stamp (in seconds), return null if fail.
+   * @group datetime_funcs
+   * @since 1.5.0
+   */
+  def unix_timestamp(s: Column, p: Column): Column = UnixTimestamp(Seq(s.expr, p.expr))
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: p should just be a String

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala (hunk 1 above)
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala]: should use `java.util.Date`, forget to change it, the construction cost could be lower than `java.sql.Date`. `java.sql.Date` is intended for Date, which is a little different for the meaning of timestamp, though it does support usage  like timestamp well.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala [hunk 4]
```diff
@@ -254,19 +255,115 @@ case class DateFormatClass(left: Expression, right: Expression) extends BinaryEx
...
+  override def dataType: DataType = LongType
+
+  override def nullSafeEval(time: Any, format: Any): Any = {
+    left.dataType match {
+      case DateType =>
+        DateTimeUtils.daysToMillis(time.asInstanceOf[Int]) / 1000L
+      case TimestampType =>
+        time.asInstanceOf[Long] / 1000000L
+      case StringType =>
+        val formatString = format.asInstanceOf[UTF8String].toString
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala]: This could be cached.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala [hunk 5]
```diff
@@ -254,19 +255,115 @@ case class DateFormatClass(left: Expression, right: Expression) extends BinaryEx
...
+    left.dataType match {
+      case StringType =>
+        val sdf = classOf[SimpleDateFormat].getName
+        nullSafeCodeGen(ctx, ev, (string, format) => {
+          s"""
+            try {
+              ${ev.primitive} =
+                (new $sdf($format.toString())).parse($string.toString()).getTime() / 1000L;
+            } catch (java.lang.Throwable e) {
+              ${ev.isNull} = true;
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala]: Should we return 0 if failed? as said in JIRA.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala (hunk 4 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala]: To simplify it, we could only cache it if `right` is foldable.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala [hunk 6]
```diff
@@ -274,4 +275,55 @@ class DateExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
       NextDay(Literal(Date.valueOf("2015-07-23")), Literal("th")),
       DateTimeUtils.fromJavaDate(Date.valueOf("2015-07-30")))
   }
+
+  test("from_unixtime") {
+    val sdf1 = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
+    val fmt2 = "yyyy-MM-dd HH:mm:ss.SSS"
```
Comment (davies) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: Could we have a test for invalid format and invalid time in String?

Comment (davies) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: Also for unix_timestamp

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 7]
```diff
@@ -2091,6 +2091,44 @@ object functions {
...
 
+  /**
+   * Gets current Unix timestamp in seconds.
+   * @group datetime_funcs
+   * @since 1.5.0
+   */
+  def from_unixtime(ut: Column): Column = FromUnixTime(ut.expr, Literal("yyyy-MM-dd HH:mm:ss"))
+
+  /**
+   * Gets current Unix timestamp in seconds.
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: This is not correct.

File: sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala [hunk 8]
```diff
@@ -206,4 +206,43 @@ class DateFunctionsSuite extends QueryTest {
...
+    val df = Seq((1000, "yyyy-MM-dd HH:mm:ss.SSS"), (-1000, "yy-MM-dd HH-mm-ss")).toDF("a", "b")
+    checkAnswer(
+      df.select(from_unixtime(col("a"))),
+      Seq(Row(sdf1.format(new Timestamp(1000000))), Row(sdf1.format(new Timestamp(-1000000)))))
+    checkAnswer(
+      df.select(from_unixtime(col("a"), fmt2)),
+      Seq(Row(sdf2.format(new Timestamp(1000000))), Row(sdf2.format(new Timestamp(-1000000)))))
+    checkAnswer(
+      df.select(from_unixtime(col("a"), fmt3)),
+      Seq(Row(sdf3.format(new Timestamp(1000000))), Row(sdf3.format(new Timestamp(-1000000)))))
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala]: Could we have a few test cases to test it as a function in SQL (using selectExpr())?

**Comment by davies** [comment]
@adrian-wang Could you also pull in some information of these functions from JIRA to PR description (for better git log)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala (hunk 5 above)
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala]: jira description comes from hive language manual, but actually hive didn't return 0 when fail.

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala]: That's weird, could doc this some where?

Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala]: done.

File: sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala (hunk 8 above)
Comment (adrian-wang) [inline sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala]: done.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala (hunk 6 above)
Comment (adrian-wang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala [hunk 9]
```diff
@@ -254,19 +255,231 @@ case class DateFormatClass(left: Expression, right: Expression) extends BinaryEx
...
+      null
+    } else {
+      left.dataType match {
+        case DateType =>
+          DateTimeUtils.daysToMillis(t.asInstanceOf[Int]) / 1000L
+        case TimestampType =>
+          t.asInstanceOf[Long] / 1000000L
+        case StringType if right.foldable =>
+          if (constFormat != null) {
+            Try(new SimpleDateFormat(constFormat.toString).parse(
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala]: We should also cache the `SimpleDateFormat`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala [hunk 10]
```diff
@@ -254,19 +255,231 @@ case class DateFormatClass(left: Expression, right: Expression) extends BinaryEx
...
+            }
+          }
+        """
+      }
+    } else {
+      nullSafeCodeGen(ctx, ev, (seconds, f) => {
+        s"""
+        try {
+          ${ev.primitive} = UTF8String.fromString((new $sdf($f.toString())).format(
+            new java.sql.Timestamp($seconds * 1000L)));
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala]: java.util.Date?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala [hunk 11]
```diff
@@ -254,19 +255,231 @@ case class DateFormatClass(left: Expression, right: Expression) extends BinaryEx
...
+        """
+      } else {
+        val t = left.gen(ctx)
+        s"""
+          ${t.code}
+          boolean ${ev.isNull} = ${t.isNull};
+          ${ctx.javaType(dataType)} ${ev.primitive} = ${ctx.defaultValue(dataType)};
+          if (!${ev.isNull}) {
+            try {
+              ${ev.primitive} = UTF8String.fromString(new $sdf("${constFormat.toString}").format(
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala]: Should we cache the `SimpleDateFormat`? see String functions as example.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala (hunk 9 above)
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala]: There are many expressions rely on `SimpleDateFormat`, including some merged expressions such as `DateFormatClass`. Let's do this later, maybe we can extract a common trait for this kind of expressions.

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala]: SGTM.

**Comment by davies** [comment]
LGTM, going to merging into master (fixing the j.s.TimeStamp), thanks!

**Comment by rxin** [comment]
@davies / @adrian-wang can one of you submit a patch to also use NonFoldableLiteral in unit tests? It's important to test the non-foldable code gen path.
